sdks/go/pkg/beam/transforms/stats/quantiles_test.go

```diff
@@ -0,0 +1,274 @@
+	reflectx.RegisterFunc(reflect.ValueOf(less).Type(), func(_ interface{}) reflectx.Func {
+		return newIntLess()
+	})
+}
+
+type intLess struct {
+	name string
+	t    reflect.Type
+}
+
+func newIntLess() *intLess {
+	return &intLess{
+		name: reflectx.FunctionName(reflect.ValueOf(less).Interface()),
+		t:    reflect.ValueOf(less).Type(),
+	}
+}
+
+func (i *intLess) Name() string {
+	return i.name
+}
+func (i *intLess) Type() reflect.Type {
+	return i.t
+}
+func (i *intLess) Call(args []interface{}) []interface{} {
+	return []interface{}{args[0].(int) < args[1].(int)}
+}
+
+func less(a, b int) bool {
+	return a < b
+}
+
+func TestLargeQuantiles(t *testing.T) {
+	const numElements int = 30000
+	inputSlice := make([]int, 0, numElements)
+	for i := 0; i < numElements; i++ {
+		inputSlice = append(inputSlice, i)
+	}
+	p, s, input, expected := ptest.CreateList2(inputSlice, [][]int{[]int{10006, 19973}})
+	quantiles := ApproximateQuantiles(s, input, less, Opts{
+		K:            200,
+		NumQuantiles: 3,
+	})
+	passert.Equals(s, quantiles, expected)
+	if err := ptest.Run(p); err != nil {
+		t.Errorf("ApproximateQuantiles failed: %v", err)
+	}
+}
+
+func TestLargeQuantilesReversed(t *testing.T) {
+	const numElements int = 30000
+	inputSlice := make([]int, 0, numElements)
+	for i := numElements - 1; i >= 0; i-- {
+		inputSlice = append(inputSlice, i)
+	}
+	p, s, input, expected := ptest.CreateList2(inputSlice, [][]int{[]int{9985, 19959}})
+	quantiles := ApproximateQuantiles(s, input, less, Opts{
+		K:            200,
+		NumQuantiles: 3,
+	})
+	passert.Equals(s, quantiles, expected)
+	if err := ptest.Run(p); err != nil {
+		t.Errorf("ApproximateQuantiles failed: %v", err)
+	}
+}
+
+func TestBasicQuantiles(t *testing.T) {
+	// Test asking for 3 quantiles for k=3.
+	tests := []struct {
+		Input    []int
+		Expected [][]int
+	}{
+		{[]int{}, [][]int{}},
+		{[]int{1}, [][]int{[]int{1}}},
+		{[]int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20}, [][]int{[]int{6, 13}}},
+	}
+
+	for _, test := range tests {
+		p, s, in, exp := ptest.CreateList2(test.Input, test.Expected)
+		quantiles := ApproximateQuantiles(s, in, less, Opts{
+			K:            3,
+			NumQuantiles: 3,
+		})
+		passert.Equals(s, quantiles, exp)
+
+		if err := ptest.Run(p); err != nil {
+			t.Errorf("ApproximateQuantiles(%v) != %v: %v", test.Input, test.Expected, err)
+		}
+	}
+}
+
+func weightedElementToKv(e testWeightedElement) (int, int) {
+	return e.Weight, e.Element
+}
+
+type testWeightedElement struct {
+	Weight  int
+	Element int
+}
+
+func TestWeightedQuantiles(t *testing.T) {
+	// Test asking for 3 quantiles for k=3.
+	input := []testWeightedElement{
+		{Weight: 1, Element: 1},
+		{Weight: 10, Element: 2},
+		{Weight: 1, Element: 3},
+		{Weight: 10, Element: 4}}
+	expected := []int{2, 4}
+	p, s, in, exp := ptest.CreateList2(input, [][]int{expected})
+	elementWeightKvs := beam.ParDo(s, weightedElementToKv, in, beam.TypeDefinition{Var: beam.TType, T: reflect.TypeOf(1)})
```

Comment:
```suggestion
	elementWeightKvs := beam.ParDo(s, weightedElementToKv, in, beam.TypeDefinition{Var: beam.TType, T: reflect.TypeOf(int(1))})
```
Numeric constants are untyped in Go, so the actual type of 1 varies by context. In this case, it'll be int. It's a preference to make this explicit.

Alternatively, since you've got reflectx imported, we already have an explict reflect.Type for it that can be used instead.

```suggestion
	elementWeightKvs := beam.ParDo(s, weightedElementToKv, in, beam.TypeDefinition{Var: beam.TType, T: reflectx.Int})
```